File: backend/src/agents/judicial/result-reviewer.ts
import { RulesEngine } from './rules-engine';
import { ExecutionReport } from '../../schemas/act';
import { ResultReviewResult } from '../../schemas/verdict';

export class ResultReviewer {
  private _rules: RulesEngine;

  constructor(rules: RulesEngine) {
    this._rules = rules;
  }

  public async reviewDelivery(
    petition: string,
    executionReport: ExecutionReport
  ): Promise<ResultReviewResult> {
    const outputs = executionReport.task_results
      .filter(r => r.status === 'success' && r.output)
      .map(r => r.output);
    
    // Bug 5 fix: 当无有效产出时，直接判定为最大偏离，而不是把
    // 字面量 '(无有效产出)' 传给 LLM scorer（会产生随机分数）
    if (outputs.length === 0) {
      return {
        deviation: {
          score: 1.0,
          explanation: '所有执行任务均失败，无有效产出。偏离度强制设为 1.0。',
          passed: false,
        },
        passed: false,
      };
    }

    const combinedOutput = outputs.join('\n');

    const deviation = await this._rules.checkDeviation(petition, combinedOutput);
    
    return {
      deviation,
      passed: deviation.passed,
    };
  }
}

File: openclaw_republic/agents/judicial/result_reviewer.py
"""结果违宪审查 — 交付前验收。

比对选民原始请愿与最终执行产出，通过 LLM-as-a-Judge
评估偏离度以检测幻觉或目标偏移。
"""

from __future__ import annotations

from openclaw_republic.agents.judicial.rules_engine import RulesEngine
from openclaw_republic.schemas.act import ExecutionReport
from openclaw_republic.schemas.verdict import ResultReviewResult


class ResultReviewer:
    """结果违宪审查 — 交付前验收。"""

    def __init__(self, rules: RulesEngine) -> None:
        """初始化结果审查器。

        Args:
            rules: 规则引擎实例。
        """
        self._rules = rules

    async def review_delivery(
        self,
        petition: str,
        execution_report: ExecutionReport,
    ) -> ResultReviewResult:
        """审查执行结果是否偏离原始请愿。

        从执行报告中提取产出文本，调用规则引擎的偏离度检查。

        Args:
            petition: 选民原始请愿文本。
            execution_report: 行政分支的执行报告。

        Returns:
            结果审查汇总。
        """
        # 从所有成功步骤中拼接产出
        outputs = [
            r.output
            for r in execution_report.task_results
            if r.status == "success" and r.output
        ]
        combined_output = "\n".join(outputs) if outputs else "(无有效产出)"

        deviation = await self._rules.check_deviation(petition, combined_output)
        return ResultReviewResult(
            deviation=deviation,
            passed=deviation.passed,
        )
